File: design/problem_book/drinks-selling-machine/Python/extensions/ext4/drink_selling_machine.py
from base.drink_selling_machine.drink_selling_machine import DrinkSellingMachine


class DrinkSellingMachineWithStats(DrinkSellingMachine):
    def __init__(self, component_cls=None):
        super().__init__(component_cls=component_cls)
        self._stats = {'drink_stats': {drink_name: 0 for drink_name in super().get_menu()},
                       'client_stats': 0}

    def get_menu(self):
        self._stats['client_stats'] = self._stats['client_stats'] + 1
        res = super().get_menu()
        return res

    def make_order(self, ordered_position):
        res = super().make_order(ordered_position)
        if res:
            self._stats['drink_stats'][ordered_position] += 1
        return res

    def print_statistics(self):
        print('Sold drinks:')
        for i in self._stats['drink_stats']:
            s = self._stats['drink_stats'][i]
            print(f'Drink name: {i}, amount of sold positions: {s}')
        k = self._stats['client_stats']
        print(f'Amount of clients: {k}')

File: design/problem_book/drinks-selling-machine/C++/extensions/ext4/DrinkSellingMachine.cpp
#include "DrinkSellingMachine.h"

#include <iostream>

DrinkSellingMachineWithStats::Menu DrinkSellingMachineWithStats::getMenu()
{
	m_clientCounter++;
	return DrinkSellingMachine::getMenu();
}

const Item* DrinkSellingMachineWithStats::makeOrder(const std::string& drink)
{
	m_drinkStats[drink]++;
	return DrinkSellingMachine::makeOrder(drink);
}

void DrinkSellingMachineWithStats::printStatistics()
{
	std::cout << "Sold stats\n";
	for (const auto& d : m_drinkStats)
		std::cout << "Drink name: " << d.first << " amount of sold positions: " << d.second << std::endl;

	std::cout << "Amount of clients: " << m_clientCounter << std::endl;
}
